### PR TITLE
Upgrade and reenable tree-sitter-haskell

### PR DESCRIPTION
**Pull request by 414owen**
I've made changes to `tree-sitter-haskell` that remove the need for c++14. It now happily compiles with c++11.
I think this is ubiquitous enough to reenable it by default?

### CODE REVIEW

**Comment by kirawi** [comment]
Looks like you'll have to enable C++11 in `build_library()`.

**Comment by archseer** [comment]
https://github.com/helix-editor/helix/blob/0dab6c8c17c027092e117a48f60feb099cd8d59c/helix-syntax/build.rs#L76

Should look something like https://github.com/helix-editor/helix/commit/212f6bc372c4ab1f4f8a3a6fe6948cddefe2fda1

**Comment by kirawi** [comment]
I see you also have a PR to drop the requirement to C++03: https://github.com/414owen/tree-sitter-haskell/pull/1
I'm not sure, but I think you wouldn't have to change anything else in Helix if that were to get merged,

**Comment by 414owen** [comment]
> I see you also have a PR to drop the requirement to C++03: [414owen/tree-sitter-haskell#1](https://github.com/414owen/tree-sitter-haskell/pull/1) I'm not sure, but I think you wouldn't have to change anything else in Helix if that were to get merged,

I'll see if I can get that touched up.

**Comment by 414owen** [comment]
Okay I've changed `tree-sitter-haskell` to only rely on a c99 compiler. Should be portable enough :)

**Comment by archseer** [comment]
This is great!

You will probably also want to pull in the `locals.scm` query and update the `highlights.scm` one: https://github.com/tree-sitter/tree-sitter-haskell/tree/master/queries

The current file predates the tree-sitter-haskell rewrite.

**Comment by 414owen** [comment]
@archseer Looks like those files were updated 19 days ago and 7 months ago, whereas the `haskell-tree-sitter` grammar's last major features were added in March 2021. I think this should be okay?

It certainly seems to be highlighting correctly when I test it.

If not, what's the process here? Cross-reference helix's known token categories with what `haskell-tree-sitter` produces?

**Comment by kirawi** [comment]
You should be able to copy the ones in the repo, and then what you'll have to do is change the captures (e.g. `@number`) to [the ones used by Helix](https://docs.helix-editor.com/themes.html#syntax-highlighting) (e.g. `@constant.numeric.integer`).

**Comment by 414owen** [comment]
Okay everything in `highlights.scm` is now understood by helix.

**Comment by 414owen** [comment]
The worst mapping (preserved from before) is:

```
[
  "forall"
  "∀"
] @keyword.control.repeat
```

In haskell these are quantifiers, and have nothing to do with loops. I don't suppose `helix` wants to add a `quantifier` scope?

**Comment by archseer** [comment]
Do you want to style these separately? If not, it's enough to use `@keyword`

File: runtime/queries/haskell/highlights.scm [hunk 1]
```diff
@@ -1,45 +1,125 @@
...
-(signature name: (variable) @type)
-(function name: (variable) @function)
-(constraint class: (class_name (type)) @class)
-(class (class_head class: (class_name (type)) @class))
-(instance (instance_head class: (class_name (type)) @class))
+;; ----------------------------------------------------------------------------
+;; Literals and comments
+
 (integer) @constant.numeric.integer
+(exp_negation) @constnats.numeric.integer
```
Comment (archseer) [inline runtime/queries/haskell/highlights.scm:5]:
```suggestion
(exp_negation) @constant.numeric.integer
```

File: runtime/queries/haskell/highlights.scm [hunk 2]
```diff
@@ -1,45 +1,125 @@
...
-"deriving" @keyword
-"via" @keyword
-"stock" @keyword
-"anyclass" @keyword
-"do" @keyword
-"mdo" @keyword
-"rec" @keyword
+(string) @string
+
+(con_unit) @symbol  ; unit, as in ()
```
Comment (archseer) [inline runtime/queries/haskell/highlights.scm:10]: Maybe `@constant.builtin` or `@constructor`? There's `@string.special.symbol` but that doesn't quite fit here

**Comment by 414owen** [comment]
> Do you want to style these separately? If not, it's enough to use `@keyword`

No that's okay. I've changed them to `keyword`.

**Comment by 414owen** [comment]
One last update to `highlights.scm` to add infix operators (recently added to `tree-sitter-haskell`).

Think we're ready to merge?

**Review by archseer — APPROVED** [review]
Thanks for working on this! 🎉

**Pull request merged**